src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -131,13 +201,128 @@ private static bool ApplyChangesFromTC25ThroughTC28(JObject sarifLog)
+            fullyQualifiedLogicalNameToIndexMap = new Dictionary<string, int>();
+
+            foreach (JProperty logicalLocationEntry in logicalLocations.Properties())
+            {
+                // This condition may occur if we've already procssed a parent logical
+                // location that happened to be enumerated after one of its children
+                if (fullyQualifiedLogicalNameToIndexMap.ContainsKey(logicalLocationEntry.Name)) { continue; }
+
+                fullyQualifiedLogicalNameToIndexMap = CreateLogicalLocationArrayEntry(
+                    logicalLocations,
+                    logicalLocationEntry.Name, 
+                    (JObject)logicalLocationEntry.Value,
+                    logicalLocationToIndexMap, 
+                    jObjectToIndexMap,
+                    fullyQualifiedLogicalNameToIndexMap);
+            }
+
+            var updatedArrayElements = new JObject[jObjectToIndexMap.Count];
+
+            foreach (KeyValuePair<JObject, int> keyValuePair in jObjectToIndexMap)
+            {
+                int index = keyValuePair.Value;
+                JObject updatedLogicalLocation = keyValuePair.Key;
+                updatedArrayElements[index] = updatedLogicalLocation;
+            }
+
+            return new JArray(updatedArrayElements);
+        }
+
+        private static Dictionary<string, int> CreateLogicalLocationArrayEntry(
+            JObject logicalLocationsDictionary,
+            string keyName,
+            JObject logicalLocation,
+            Dictionary<LogicalLocation, int> logicalLocationToIndexMap,
+            Dictionary<JObject, int> jObjectToIndexMap,
+            Dictionary<string, int> keyToIndexMap)
+        {
+            int index;
+            keyToIndexMap = keyToIndexMap ?? new Dictionary<string, int>();
+
+            string fullyQualifiedName = keyName;
+
+            // Parent key is no longer relevant and will be removed.
+            // We require 
+            string parentKey = (string)logicalLocation["parentKey"];
+
+            if (parentKey == null)
+            {
+                // No parent key? We are a root location. 
+                logicalLocation["parentIndex"] = -1;
+            }
+            else
+            {
+                // Have we processed our parent yet? If so, retrieve the parentIndex and we're done
+                if (!keyToIndexMap.TryGetValue(parentKey, out index))
+                {
+                    // The parent hasn't been processed yet. We must force its creation
+                    // determine its index in our array. This code path results in 
+                    // an array order that does not precisely match the enumeration
+                    // order of the logical locations dictionary.
+                    keyToIndexMap = CreateLogicalLocationArrayEntry(
+                        logicalLocationsDictionary,
+                        parentKey,
+                        (JObject)logicalLocationsDictionary[parentKey],
```

Comment:
>parentKey [](start = 60, length = 9)

If the SARIF file is invalid, in that it contains a `logicalLocation` object with a `parentKey`, but the parent isn't in the dictionary, then you'll get some generic "Key not found" exception. Consider checking for this, and throwing an exception with a message like "The logical location 'X::Y' specifies that its parent location is 'X', but the `run.logicalLocations` dictionary does not have an entry with key 'X'.".

Reply:
Actually, the semantics of this JSON dictionary is that a null value will be returned. I can easily add a null check at the beginning of the method.

---
In reply to: [243857356](https://github.com/Microsoft/sarif-sdk/pull/1170#discussion_r243857356) [](ancestors = 243857356)